### PR TITLE
add main litentry address into IDGraph

### LINKED ISSUE #1174: Should the IDGraph contain the main litentry identity?
### Context
It will affect the identity vector that is passed into the assertion building.

Or is it assertion specific? Like when calculating the total amount of LIT tokens held, we should definitely include it.

Please share your opinions too @jonalvarezz @kaylawangnow

---
:heavy_check_mark: Please set appropriate **labels** and **assignees** if applicable.

### PR DESCRIPTION
**Pull request by BillyWooo**
Add one special item when main litentry account not existed. 

Resolved #1174 

```
if IDGraphs::<T>::get(&who, &prime_id).is_none() {
    // Not existed, so create the prime entry.
    let context = IdentityContext::<T> {
	    metadata: None,
	    creation_request_block: Some(0),
	    verification_request_block: Some(0),
	    is_verified: false,
    };
    IDGraphs::<T>::insert(&who, &prime_id, context);
}
```

### CODE REVIEW

File: tee-worker/litentry/pallets/identity-management/src/lib.rs [hunk 1]
```diff
@@ -201,16 +201,32 @@ pub mod pallet {
...
+				network: SubstrateNetwork::Litentry,
+				address: prime_user_address,
+			};
+			if IDGraphs::<T>::get(&who, &prime_id).is_none() {
+				// Not existed, so create the prime entry.
+				let context = IdentityContext::<T> {
+					metadata: None,
+					creation_request_block: Some(0),
+					verification_request_block: Some(0),
+					is_verified: false,
```
Comment (Kailai-Wang) [inline tee-worker/litentry/pallets/identity-management/src/lib.rs]: I think `is_verified` should be true

**Review by Kailai-Wang — COMMENTED** [review]
Thanks. I think we still miss a few things:
- IMP mock needs to be adjusted
- we should disallow deletion of prime_identity in `remove_identity`
- same applies to `verify_identity` => we might need an error like `IdentityAlreadyVerified`
- ts-tests should test about the size and content of the ID-graph (to make sure we do have the prime id)

**Review by jingleizhang — CHANGES_REQUESTED** [review]
Please add a case in ts-test, to cover the `RemovePrimeIdentityDisallowed` error. Thanks!

**Review by jingleizhang — APPROVED** [review]

**Comment by jingleizhang** [comment]
According to failed CI [log](https://github.com/litentry/litentry-parachain/actions/runs/4364579438/jobs/7634339978), there is a panic:

> [2023-03-08T15:17:42Z DEBUG itc_parentchain_indirect_calls_executor] Scanning block 77 for relevant xt
thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: Error { cause: None, desc: "Not enough data to decode vector" }', /root/work/tee-worker/core/parentchain/indirect-calls-executor/src/executor/litentry/batch_all.rs:343:81
stack backtrace:
   0: rust_begin_unwind
   1: core::panicking::panic_fmt
             at library/core/src/panicking.rs:65
   2: core::result::unwrap_failed
             at library/core/src/result.rs:1791
   3: <core::slice::iter::IterMut<T> as core::iter::traits::iterator::Iterator>::for_each
   4: <E as itc_parentchain_indirect_calls_executor::executor::DecorateExecutor<ShieldingKeyRepository,StfEnclaveSigner,TopPoolAuthor,NodeMetadataProvider>>::decode_and_execute
   5: <itc_parentchain_indirect_calls_executor::IndirectCallsExecutor<ShieldingKeyRepository,StfEnclaveSigner,TopPoolAuthor,NodeMetadataProvider> as itc_parentchain_indirect_calls_executor::ExecuteIndirectCalls>::execute_indirect_calls_in_extrinsics
   6: <itc_parentchain_block_importer::block_importer::ParentchainBlockImporter<ParentchainBlock,ValidatorAccessor,StfExecutor,ExtrinsicsFactory,IndirectCallsExecutor> as itc_parentchain_block_importer::ImportParentchainBlocks>::import_parentchain_blocks
   7: <itc_parentchain_block_import_dispatcher::triggered_dispatcher::TriggeredDispatcher<BlockImporter,BlockImportQueue> as itc_parentchain_block_import_dispatcher::triggered_dispatcher::TriggerParentchainBlockImport>::import_until
   8: <T as its_consensus_slots::PerShardSlotWorkerScheduler<ParentchainBlock>>::on_slot
   9: execute_trusted_calls
  10: sgx_execute_trusted_calls
  11: do_ecall
  12: enter_enclave
  13: <unknown>
  14: <unknown>
note: Some details are omitted, call backtrace::enable_backtrace() with 'PrintFormat::Full' for a verbose backtrace.
fatal runtime error: failed to initiate panic, error 5

**Review by jingleizhang — CHANGES_REQUESTED** [review]
please fix the panic bug

**Comment by BillyWooo** [comment]
> please fix the panic bug
The panic bug was introduced by a merge from other branch. It's already fixed there and merged back.

**Review by jingleizhang — APPROVED** [review]

File: tee-worker/litentry/pallets/identity-management/src/lib.rs [hunk 2]
```diff
@@ -199,18 +201,37 @@ pub mod pallet {
...
-						.try_into()
-						.map_err(|_| DispatchError::Other("invalid account id"))?;
-					let user_address: Address32 = address_raw.into();
-					ensure!(user_address != address, Error::<T>::IdentityShouldBeDisallowed);
+					ensure!(prime_user_address != address, Error::<T>::IdentityShouldBeDisallowed);
 				}
 			}
+
+			let prime_id = Identity::Substrate {
+				network: SubstrateNetwork::Litentry,
```
Comment (Kailai-Wang) [inline tee-worker/litentry/pallets/identity-management/src/lib.rs:221]: This won't work if we are on other mainnets (e.g. rococo)

The network has to be created from `parent_ss58_prefix` - alternatively we can still hardcode it but need to change it before we deploy on other chains

File: tee-worker/litentry/pallets/identity-management/src/lib.rs [hunk 3]
```diff
@@ -247,6 +284,13 @@ pub mod pallet {
 			IDGraphs::<T>::try_mutate(&who, &identity, |context| -> DispatchResult {
 				let mut c = context.take().ok_or(Error::<T>::IdentityNotExist)?;
 
+				if c.metadata.is_none()
+					&& c.creation_request_block == Some(0)
+					&& c.verification_request_block == Some(0)
+				{
+					ensure!(!c.is_verified, Error::<T>::IdentityAlreadyVerified);
+				}
+
```
Comment (Kailai-Wang) [inline tee-worker/litentry/pallets/identity-management/src/lib.rs:293]: I think this should be extended to non-prime Identities as well - we should quite early if it's already verified.

File: tee-worker/litentry/pallets/identity-management/src/lib.rs [hunk 4]
```diff
@@ -199,18 +201,37 @@ pub mod pallet {
...
+				ensure!(
+					!(c.is_verified && c.creation_request_block != Some(0)),
+					Error::<T>::IdentityAlreadyVerified
+				);
 			}
+			let prime_address_raw: [u8; 32] = who
+				.encode()
+				.try_into()
+				.map_err(|_| DispatchError::Other("invalid account id"))?;
+			let prime_user_address: Address32 = prime_address_raw.into();
```
Comment (Kailai-Wang) [inline tee-worker/litentry/pallets/identity-management/src/lib.rs:213]: Can't we directly compare `prime_id` and `identity`?

File: tee-worker/litentry/pallets/identity-management/src/lib.rs [hunk 5]
```diff
@@ -230,6 +251,22 @@ pub mod pallet {
...
+			{
+				if metadata.is_none()
+					&& creation_request_block == Some(0)
+					&& verification_request_block == Some(0)
+					&& is_verified
+				{
+					ensure!(false, Error::<T>::RemovePrimeIdentityDisallowed);
+				}
+			}
+
```
Comment (Kailai-Wang) [inline tee-worker/litentry/pallets/identity-management/src/lib.rs:269]: Same as above - I think it's enough to check if `identity` is `prime_id`

**Review by Kailai-Wang — COMMENTED** [review]
A few hiccups I found

**Pull request merged**